### ISSUE TITLE
Load balancer behavior results in high SG memory utilization

### ISSUE THREAD

**Issue by zgramana** [couchbase/sync_gateway#768]
Created to track [this issue raised by Brett Harrison on the mailing list](https://groups.google.com/d/msgid/mobile-couchbase/34cd2982-9d26-47c2-8ffd-85faa39acd7c%40googlegroups.com?utm_medium=email&utm_source=footer):

> I am seeing a memory leaks in sync_gateway (1.0.4)
> 
> I am not a Go programmer but I was hoping to gather some data that would be useful in finding these leaks. 
> I turned on profiling and took heap data snapshots about 12 hours apart.  I hope I did this correctly.
> Is this data useful in finding the leaks?  Is there anything else I could do to help?

**Comment by zgramana** [comment]
Can you provide SG log data from the crash? A stack trace would be very helpful.

In line with Nico’s observation in issue #509, do you have any documents with a very high number of revisions?

Also, I didn’t see if you mentioned below which OS you are on.

**Comment by brettharrisonzya** [comment]
I am running ubuntu 14.4.  We don't have a very high amount of revisions.  Some have 24 revisions at most.  I'll try to catch the next log when it crashes.

**Comment by zgramana** [comment]
Thanks

**Comment by adamcfraser** [comment]
Relevant details from the mailing list thread to try to reproduce this:
- only about 35 docs in the DB, all in channel "public"
- clients all replicate using the guest user
- clients are pull-only; no updates coming from clients

@brettharrisonzya - what's the replication mode from the client?  You mentioned you recently stopped using continuous replication for clients - what's it doing now?

**Comment by brettharrisonzya** [comment]
Currently I do a pull at app startup and again if the app resumes.
Using the .net client I set "pull.Continuous = false;" and then call "pull.Start();" on those events.

**Comment by brettharrisonzya** [comment]
I'm waiting for this to crash.  But here is todays memory usage...

top - 00:54:53 up 8 days,  8:11,  1 user,  load average: 0.00, 0.02, 0.05
Tasks:  98 total,   2 running,  96 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:   6153524 total,  6105860 used,    47664 free,     2060 buffers
KiB Swap:  2096444 total,   600752 used,  1495692 free.    66820 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND  
  538 root      20   0 7166784 5.655g   3692 S   0.3 96.4 501:34.27 sync_gateway

**Comment by ajres** [comment]
I'm trying to produce a test for this issue, @brettharrisonzya you mentioned that your docs have attachments, what size are the attached files?

**Comment by brettharrisonzya** [comment]
There are 2 attachments on most of the docs.  The first averages around 50kb and the second around 100kb.

**Comment by ajres** [comment]
Items to complete
- [x]  Gatling test scenario
- [x]  Run against local SG on MAC OS X (not reproduced)
-  [x] Run against Perfcluster in AWS

**Comment by zgramana** [comment]
@brettharrisonzya can you confirm what load balancers you have deployed in front of SG, if any? If so, please add make/model/version (or if in a public cloud, as much info as you have). Cheers.

**Comment by ajres** [comment]
So far unable to generate an unbounded memory leak using the parameters listed in this ticket. Trying additional variations with blocking longpoll _changes in #875

**Comment by brettharrisonzya** [comment]
We are using haproxy 1.5.8 on ubuntu as the load balancer.

Since I have changed our app to only pull changes on startup / resume the issue is not as pronounced as it was when replication was set to continuous.

**Comment by zgramana** [comment]
To @brettharrisonzya, @jonlipsky, @vladoatanasov, and anyone else who has (or knows someone who has) experienced this issue, please help us reproduce this issue by completing this questionnaire: http://goo.gl/forms/2exqyyY2kP. Cheers!

**Comment by zgramana** [comment]
The issue appears to be a complex interaction between SG, chunked transfer encoding, gzip compression, the load balancer, and Couchbase Lite for Android and .NET.

In short, CBL Android and .NET were not properly modifying the heartbeat interval in response to proxies that aggressively closed "idle" connections. The use of chunked + gzip often causes proxies to buffer the entire http response, and since the SG uses the chunk delimiter as the heartbeat value, proxies often just keep eating the packets until their buffers fill, or they decided it's idle enough. 

Meanwhile, the CBLite instances are creating new HTTP requests for the changes feed. This results in SG continually allocating TCP/socket buffers for the new connections, which can eventually lead to an OOM panic with a large number of CBL-Android and CBL-NET users with active continuous replications. This problem was exacerbated in by a regression in SG from 1.0.3 -> 1.0.4 where HTTP POST for the changes feed effectively ignored the heartbeat parameter (HTTP GETs worked fine). As an additional result, Couchbase Lite for iOS would eventually stop replicating in these scenarios, after a given number of retries to find a better heartbeat value.

tl;dr: CBL Android and .NET now use an adaptive behavior in 1.1, and disabling compression between the proxy and SG, by adding `"compressResponses": false` to your sg.config, will generally resolve the issue. Also, configure your proxy to not buffer chunked transfer encoded responses.

**Comment by ajres** [comment]
Adding a new test scenario, that uses a third party load balancer to replicate the above scenario specifically connection termination and chunk encoding/gzip compression.

Initial implementation will use nginx as the load balancer with possibility of adding HAProxy later.

We need to optionally support an load balancer in the Perfcluster-AWS for the following test scenarios.

1 - to reproduce the increasing memory scenario, similar to what has been done with Gatling client terminating the connections.  We don't expect results significantly different than the Gatling setup, but would be more definitive confirmation of a connection issue.

2 - to be able to test whether _changes heartbeats get swallowed when using compression and chunked transfer encoding, and then test that setting no-cache on the Sync Gateway _changes requests fixes that issue

**Comment by ajres** [comment]
- [x] Update perfcluster-aws cloudformation_template.py to create optional loadbalancer nodes.
- [x] Add new ansible script to install and configure nginx as a round robin load balancer in front of Sync Gateway nodes.  Configure connection timeouts. 
- [x] Add a new ansible script to configure gatling to connect to load balancer node(s) rather than the Sync Gateway nodes. 
- [x] Update the gatling connection test to not terminate connections in the gatling client but allow them to be terminated in nginx.
- [x] Add nginx plugin to support chunked encoding and and configure to simulate customer deployments.
- [ ] Document how to run this test on customer premises outside of the perfcluster-aws environment.

**Comment by hegra** [comment]
We have a similar issue.
A few hours of running sync gateway eats 3 GB of memory and a restart of sync gateway is needed.

Running the following version: https://aws.amazon.com/marketplace/pp/B00FA8DO50 on single m1.medium instance. Also upgraded sync gateway to master(bfb402f) but no noticeable difference.
No proxy in use that I'm aware of.

5 android (couchbase-lite-android:1.1.0-532) clients connected with continuous sync in push mode. Documents of around 100kb, no attachments. Documents updated a few times every minute so revisions can get quite long. Using revs_limit:20

"compressResponses": false in config.json didn't have any effect, if this is the right place to add it?

Update: The following client code can consume the memory within 30 min: http://pastebin.com/3mBsENy1  (Need to set sleep to around 1000 ms for every document else a memory leak in the client code will be exposed before!)

**Comment by adamcfraser** [comment]
@hegra That sounds like an interesting case, particularly if you aren't doing any pull replication.  If you can complete the questionnaire at http://goo.gl/forms/2exqyyY2kP, it will help us reproduce this scenario.

**Comment by hegra** [comment]
I have completed the form and the memory leak is quite severe as it looks now in new tests. We need to restart sync gateway every second hour. Since we are running on the default aws couchbase ami it should be possible to reproduce.

**Comment by adamcfraser** [comment]
@hegra It's possible that the memory usage you're seeing is related to the revision caching Sync Gateway does.  Do your doucments that are getting updated a few times a minute typically have conflicting updates?

**Comment by ajres** [comment]
Waiting for results of running longpoll _changes requests running againt customer cluster. 

Writing a new test to reproduce push only scenario reported by @hegra

**Comment by hegra** [comment]
@adamcfraser Sometimes conflicts occurs in normal operation, but did a test now with a fresh bucket and only one client that updates the document locally once a second and then no conflicts occur (At least not displayed in the log) but memory is still consumed fast. 
Also tested with walrus now and same behavior.

The following is repeated in the log:
2015-06-19T17:25:15.973Z HTTP:  # 327: PUT /sync_gateway/_local/5292266020ca7f9b885360e38c1f2d8aee601ef5  (as ***_)
2015-06-19T17:25:16.367Z HTTP:  # 328: POST /sync_gateway/_revs_diff  (as ***_)
2015-06-19T17:25:17.542Z HTTP:  # 329: POST /sync_gateway/_bulk_docs  (as ***_)
2015-06-19T17:25:17.559Z HTTP:  # 330: POST /sync_gateway/_revs_diff  (as ***_)
2015-06-19T17:25:17.728Z HTTP:  # 331: POST /sync_gateway/_bulk_docs  (as ***_)
2015-06-19T17:25:19.244Z HTTP:  # 332: POST /sync_gateway/_bulk_docs  (as ***_)
2015-06-19T17:25:19.260Z HTTP:  # 333: POST /sync_gateway/_revs_diff  (as ***_)
2015-06-19T17:25:20.865Z HTTP:  # 334: POST /sync_gateway/_bulk_docs  (as ***_)
2015-06-19T17:25:20.876Z HTTP:  # 335: POST /sync_gateway/_revs_diff  (as ****)

**Comment by adamcfraser** [comment]
@hegra Thanks for the details.  If this actually related to the revision caching, I'd expect the memory usage to eventually plateau.   We're doing testing now to reproduce this ourselves.

In the meantime, you could also try setting `revs_limit` for the db in your Sync Gateway config, to have Sync Gateway prune the revision tree more aggressively (the default revs limit is 1000).  This should help the memory usage of the revision cache to plateau at a lower level.

**Comment by ajres** [comment]
Revision cache issues for high write load has been moved to #936

**Comment by zgramana** [comment]
This is now a tracking ticket, and @ajres will update additional additional issue links.

**Comment by ajres** [comment]
Issues being tracked by this umbrella ticket:

#945 - Set no-caching headers on _changes HTTP response
#904 - Support for max heartbeat in Sync Gateway
#865 - Create blocking long poll gatling test
